### PR TITLE
Internal evaluations

### PR DESCRIPTION
**Pull request by entantoencuanto**
#### :tophat: What? Why?
This PR adds the view and the actions to manage internal evaluations in proposals.

#### :pushpin: Related Issues
- Fixes
  - #1 
  - #2 
  - #3 
  - #4 
  - #5

#### Testing

- In https://decidim-lot2.populate.tools/admin/participatory_processes/branch-highway/components/4/manage/proposals/2
   - See the block of Internal evaluations
   - If assigned, create/edit an evaluation
   - See created evaluations

- In https://decidim-lot2.populate.tools/admin/participatory_processes/branch-highway/components/4/manage/
  - See the new columns with Evaluation status

### :camera: Screenshots

* Adds the table with the list of assigned valuators to a proposal, including buttons to unnassign valuators (only available for admins) and edit evaluation (only available to assigned valuators for their own evaluation). The table allows users to sort by last edition and status (#2)
image

* Implements a modal to create or edit evaluations. The modal allows valuators to select a status among the statuses defined for the component and edit the evaluation text (#2)
image

* Changes the valuation assignments column in proposals index to display the status of each proposal showing the evaluations count versus the total valuations assignments (#4)
Screenshot from 2024-07-09 11-47-33

* Allows exporting the internal evaluations (#5)
Screenshot 2024-07-09 at 12 05 06

* Defines internal evaluations factory and adds some tests

:hearts: Thank you!

### CODE REVIEW

**Comment by furilo** [comment]
We have done all the issues in a single PR because they where very dependant.

**Review by furilo — APPROVED** [review]

**Comment by entantoencuanto** [comment]
Great job, @mllocs! I have reviewed the serialization part and it works perfectly

**Comment by NilHomedes** [comment]
Hi @entantoencuanto ! Here is my review, sorry it takes longer than expected.

**New feature: Internal evaluation module**
- [x] Design: Reduce “Actions” column width and increase “Evaluation” column width. 
The “Actions” column is taking too much width.
- [x] Design: Add borders to the table. 
- [x] If I fill the internal evaluation modal in a language that is not the primary, it generates an error that is not very clear. To simplify and avoid this possible error, could we make this text field without language options? 
Screenshot 2024-07-12 at 10-39-14 Review internal-evaulation module

- [x] I can delete a proposal status that has been used in an internal evaluation. If I do so, when I access to the admin of that proposal, then I get a 500 error. See: https://decidim-lot2.populate.tools/admin/participatory_processes/branch-highway/components/4/manage/proposals/3. It should not be possible to delete a status that has been used in an internal evaluation. 

**Differentiate between the proposals I have already evaluated and those pending**
- [x] As an admin, I can not filter by Evaluation. See the first AC of the issue.
Given that I’m an administrator
When I go to the proposals list
Then I can filter by “Evaluation” > “Evaluated”, “Not evaluated”

**Allow exporting internal evaluations**
- [x] Export is not working. I’m not receiving the email.

**Change evaluators column in admin proposals list**
Works fine!

**Comment by entantoencuanto** [comment]
> **Differentiate between the proposals I have already evaluated and those pending**
> 
>  * [ ]  As an admin, I can not filter by Evaluation. See the first AC of the issue.
>       Given that I’m an administrator
>       When I go to the proposals list
>       Then I can filter by “Evaluation” > “Evaluated”, “Not evaluated”
> 
How can an administrator differentiate between evaluated and not evaluated proposals? As a valuator, the criteria are clear, among the assigned proposals the distinction is made between those that she has evaluated and those that she has not.

But as an admin with no valuator role and no assigned proposals when do we consider that a proposal is evaluated? When the proposal has at least one evaluation? When all assigned evaluations are completed?

I have added the filter to appear if a user is valuator (and the filter affects to their evaluations) or if the user o admin selects a valuator in the corresponding filter
image

**Comment by NilHomedes** [comment]
> I have added the filter to appear if a user is valuator (and the filter affects to their evaluations) or if the user o admin selects a valuator in the corresponding filter

Oh, sorry @entantoencuanto I haven't noticed that. In that case, the need is covered. So an admin can select a valuator and, then, filter by "Evaluated", so the admin can easily see the proposals pending to evaluate that a valuator has.

**Comment by entantoencuanto** [comment]
Hi, @NilHomedes I have checked again the export and it seems to work  with my user. ¿Which email has the user you are using to export the evaluations?

**Comment by NilHomedes** [comment]
> Hi, @NilHomedes I have checked again the export and it seems to work with my user. ¿Which email has the user you are using to export the evaluations?

Hey @entantoencuanto, I'm not receiving it yet. I've tried with two different mails. In fact, I'm receiving the "Proposals CSV export file", but not the "Internal evaluations" file. See:
Screenshot 2024-07-16 at 09-44-08 Your export proposals-2024-07-16-27836 csv is ready - nil@decidim org - Correo de Decidim

**Comment by entantoencuanto** [comment]
Some changes added:
* Change of body to be a text without translations
* Avoid exceptions showing evaluations when the state has been deleted. I think that this change fixes the errors exporting internal evaluations also, can you check again @NilHomedes? 
* Prevent deletion of states when there are associated evaluations
* Delete evaluations associated to valuation assignments when the assignment is deleted
* Changes requested on the design of the table
* Tests additions

**Review by NilHomedes — APPROVED** [review]
Hi @entantoencuanto  everything working fine now :sparkles: 
Thanks fot he job!

**Comment by NilHomedes** [comment]
Hey @rxnetwalker! This is ready to review by the AM Barcelona team, feel free to assign it.

**Review by fblupi — CHANGES_REQUESTED** [review]
Code-wise, everything seems good. As the business logic was reviewed by Nil, I didn't take too much time on it.

I would like to propose including a test to make it easier to upgrade the decidim gem version. In other instances and modules I've been working on, we use to include a [`spec/lib/overrides_spec.rb`](https://github.com/AjuntamentdeBarcelona/decidim-module-kids/blob/main/spec/lib/overrides_spec.rb) file where we check the md5 of the files we are modifying the behavior (for example: all the views where we are using deface to change or the Proposal model where I see we are including some methods). If we upgrade the decidim version and some of those files have changed we will notice and check if those changes affect our customizations

**Comment by entantoencuanto** [comment]
Checksums checks added

File: Gemfile.lock [hunk 1]
```diff
@@ -2,78 +2,263 @@ PATH
...
   specs:
     decidim-internal_evaluation (0.28.1)
-      decidim-core (>= 0.28.1)
+      decidim-core (>= 0.28)
+      decidim-proposals (>= 0.28)
+      decidim-templates (>= 0.28)
+      deface (~> 1.9)
+
+PATH
+  remote: /home/edu/proyectos/decidim/decidim_version_movil/decidim
```
Comment (fblupi) [inline Gemfile.lock]: I was trying to create the `test_app` to run the tests locally and I realized about this. Could you change it to a version not coming from your local setup?

**Review by fblupi — CHANGES_REQUESTED** [review]

File: Gemfile.lock (hunk 1 above)
Comment (entantoencuanto) [reply Gemfile.lock]: Ah, the problem is that the module requires Decidim 0.29 release. I'm going to try to update the dependencies with the recent 0.29.0.rc1

Comment (fblupi) [reply Gemfile.lock]: You could even use 0.29.0.rc4, which seems to be the last release candidate. We could even wait for the official 0.29.0 version which will be released in a few days.

Comment (entantoencuanto) [reply Gemfile.lock]: Done! I think that the tests will run correctly in your local environment

**Review by fblupi — CHANGES_REQUESTED** [review]
Just a last small change request: Could you include `.rpsec-failures`, `spec/decidim_dummy_app/`, and `development_app/` to the `.gitignore`?

**Comment by entantoencuanto** [comment]
You are right, done!

**Review by fblupi — APPROVED** [review]

**Pull request merged**